### PR TITLE
Changes for unit cell systems - always save real-space & unit axes

### PR DESCRIPTION
**Pull request by ghutchis**
Signed-off-by: Geoff Hutchison <geoff.hutchison@gmail.com>

Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

### CODE REVIEW

**Comment by ghutchis** [comment]
This will be useful for solid-state codes. Most use real-space coordinates and translation axes, instead of cell parameters. The files are backwards-compatible, but offer additional fields on write.

Technically, this is WIP, since I haven’t added read-support - just want to make sure there are no complaints or issues before I do that.

**Comment by psavery** [comment]
Yeah, there is some benefit in explicitly providing the unit cell matrix. Some calculations do depend on the orientation of the unit cell.

But if we have the unit cell explicitly, we won't need the A, B, C, alpha, beta, gamma for the unit cell (it is information that can be derived from the unit cell matrix and therefore repetitive).

Also, I don't think we would necessarily need both cartesian and fractional coordinates, because one can convert from one to the other as well. We could just choose one to stick with.

**Comment by cryos** [comment]
I don't like the duplication aspect, but Chemical JSON is designed to have optional fields. I have been thinking about this for a while, could we pass in a JSON string with settings for readers/writers? They could be useful for commands, i.e. just give elements, give elements and proton number, explicit Cartesian, etc.

The default writer could remain pretty minimal, but some options could affect behavior.

File: avogadro/io/cjsonformat.cpp [hunk 1]
```diff
@@ -380,6 +380,20 @@ bool CjsonFormat::write(std::ostream& file, const Molecule& molecule)
...
+    vectors.push_back(molecule.unitCell()->aVector().z());
+    
+    vectors.push_back(molecule.unitCell()->bVector().x());
+    vectors.push_back(molecule.unitCell()->bVector().y());
+    vectors.push_back(molecule.unitCell()->bVector().z());
+
+    vectors.push_back(molecule.unitCell()->cVector().x());
+    vectors.push_back(molecule.unitCell()->cVector().y());
+    vectors.push_back(molecule.unitCell()->cVector().z());
+    root["cell vectors"] = vectors;
```
Comment (cryos) [inline avogadro/io/cjsonformat.cpp]: I have been using camel case for new keys, they are easier to use in all languages, i.e. cellVectors, I was hoping avoid any new keys with spaces in them.

**Comment by cryos** [comment]
Would a text container work, and for CJSON it would be JSON, for other formats it could be other things. Yes, I had been feeling the need for options for a while @ghutchis. I will see if I can put together an example, it could also be requested as a simple JSON string and the Open Babel plugin would then turn it into command line arguments, other readers unpack as they choose.

File: avogadro/io/cjsonformat.cpp (hunk 1 above)
Comment (ghutchis) [reply avogadro/io/cjsonformat.cpp]: I’m coming back to this now. I’m happy to use cellVectors as the key. I’m not sure what you mean by “vectors” for the storage? Do you want to store them as [ [ax, ay, az], [bx, by, bz], [cx, cy, cz] ] instead? That’s fine, but it’s different from how we store atomic coordinates - thus I went with an array of 9 values.

Comment (cryos) [reply avogadro/io/cjsonformat.cpp]: I meant the key name, that was all. I actually want to use this in our server platform now too. I was suggesting storing the whole thing in the "unit cell" key as "vectors" as the "unit cell" already makes it clear that it is defining the molecule's unit cell. I think the way you stored it makes sense.

Comment (ghutchis) [reply avogadro/io/cjsonformat.cpp]: Okay, I’ll get that done later this afternoon, thanks.

**Comment by ghutchis** [comment]
I’m a bit unsure on how to proceed here. If I drop the cell vectors into “unit cell” then it’s no longer a dictionary. I can do that, but it won’t be backward compatible.

I think I should just have “unit cell” contain “cellVectors” in addition to “a”, “b”, “c”, etc.

**Comment by ghutchis** [comment]
Example:
```
  "unit cell": {
    "a": 3.0,
    "alpha": 90.0,
    "b": 3.0,
    "beta": 90.0,
    "c": 3.0,
    "cellVectors": [
      3.0,
      0.0,
      0.0,
      -1.4999999999999993,
      2.5980762113533165,
      0.0,
      1.8369701987210297e-16,
      3.18172571617472e-16,
      3.0
    ],
    "gamma": 119.99999999999999
  }
```

**Comment by cryos** [comment]
> I’m a bit unsure on how to proceed here. If I drop the cell vectors into “unit cell” then it’s no longer a dictionary. I can do that, but it won’t be backward compatible.

Not sure what you mean, referring to Python dictionary? I am happy to do the work, but I think that what is produced should be set by an option, i.e. default to saving what we had before, and if the option is set then switch to Cartesian and your new cell vectors. I don't think it is useful to save both by default.

I want to start using this, and it is a relatively small change. I really want to avoid duplicating data unless that is requested in an option for the format.

**Comment by ghutchis** [comment]
Honestly, I don’t think the fractional coordinates are that useful. Is anything external using them? As far as formats, mostly CIF stores in fractional - nearly everything else uses real-space Cartesians.

What I meant by “dictionary” was the JSON - guess I should have said “object.” I definitely don’t see the point of using an option - most people will want the unit cell vectors - I had several requests at the UGM for easy access to them. I also think that dropping a, b, c, alpha, beta, gamma isn’t that big a deal - it takes up only a few extra bytes.

But if you have a better solution, by all means, it doesn’t matter much. I’d just insist that real-space Cartesians should be the default with unit cell axes. Almost all of the time, that’s what people want IMHO.

**Comment by cryos** [comment]
I don't have strong feelings on the defaults, and went with what the people who had worked with materials more in the past did. I do have strong feelings about not duplicating data, and trying to remain minimal, for larger molecules storing the Cartesian and fractional coordinates would be quite a cost.

There seems to be enough of a split that offering both would be reasonable, but switching to make Cartesian coordinates the default. We want to use this format to store data in MongoDB, and to transmit over RESTful APIs, so reducing duplication is always a concern, but we didn't relish the idea of writing TypeScript to deal with fractional coordinates.

**Review by cryos — APPROVED** [review]
I will push a follow PR later, I think we can merge what we have now, and do that later.

**Pull request merged**